packages/config/src/env/main.js

```diff
@@ -12,20 +12,48 @@ const { getGitEnv } = require('./git')
 // TODO: add `CONTEXT` and others
 const getEnv = async function ({ mode, config, siteInfo, accounts, addons, buildDir, branch, deployId, context }) {
   if (mode === 'buildbot') {
-    return { general: {}, account: {}, addons: {}, ui: {}, configFile: {}, all: {} }
+    return {}
   }
 
   const generalEnv = await getGeneralEnv({ siteInfo, buildDir, branch, deployId, context })
   const [accountEnv, addonsEnv, uiEnv, configFileEnv] = getUserEnv({ config, siteInfo, accounts, addons })
-  return {
-    general: generalEnv,
-    account: accountEnv,
-    addons: addonsEnv,
-    ui: uiEnv,
-    configFile: configFileEnv,
-    // Merging priority matters
-    all: { ...generalEnv, ...accountEnv, ...addonsEnv, ...uiEnv, ...configFileEnv },
-  }
+
+  // Sources of environment variables, in descending order of precedence.
+  const sources = [
+    { key: 'configFile', values: configFileEnv },
+    { key: 'ui', values: uiEnv },
+    { key: 'addons', values: addonsEnv },
+    { key: 'account', values: accountEnv },
+    { key: 'general', values: generalEnv },
+  ]
+
+  // A hash mapping names of environment variables to objects containing the following properties:
+  // - sources: Array of sources where the environment variable was found. The first element is the source that
+  //   actually provided the variable (i.e. the one with the highest precedence).
+  // - values: The value of the environment variable.
+  const environment = {}
+
+  sources.forEach((source) => {
+    Object.keys(source.values).forEach((key) => {
+      if (environment[key] === undefined) {
+        // Local mutation. Alternatives hinder simplicity and readibility.
+        // eslint-disable-next-line fp/no-mutation
+        environment[key] = {
+          sources: [source.key],
+          value: source.values[key],
+        }
+      } else {
+        // Local mutation. Alternatives hinder simplicity and readibility.
+        // eslint-disable-next-line fp/no-mutation
+        environment[key] = {
+          ...environment[key],
+          sources: [...environment[key].sources, source.key],
+        }
+      }
+    })
+  })
+
+  return environment
```

Comment:
[sand] Would using `map-obj` make the code simpler and remove the linting warnings? For example:

```js
const envs = [generalEnv, accountEnv, addonsEnv, uiEnv, configFileEnv]
const envNames = ['general', 'account', 'addons', 'ui', 'configFile']
return mapObj(Object.assign({}, ...envs), (name, value) => {
  const sources = envs.filter((env) => env[name] !== undefined).map((_, index) => envNames[index])
  return [name, { value, sources }]
})
```

Reply:
I wasn't too happy about disabling the linting rule, so I'm glad you brought this up.

I tried a different alternatives and wasn't convinced by any of them, but using`map-obj` is an interesting approach. Personally, I still find the mutation route easier to read, but not enough to justify disabling linting rules.

I'll change it. Thanks!

Reply:
@ehmicky You can see my changes in https://github.com/netlify/build/pull/2040/commits/a1f35168159006ac46e2c43e17c2f93377d8108c. I opted for a different approach, because the code above wasn't giving us exactly what we need, and as I added to it it started to feel a little bit too complex.

Let me know how you feel about the updated version. Happy to rejig it more if needed.

Reply:
Looks good!